src/ioc_setup.ts

```diff
@@ -0,0 +1,107 @@
+import {deferUntilFactory} from "@nicolawealth/defer_until";
+
+export const iocFactory = () => {
+  const _defaults: Record<string, unknown> = {};
+
+  // _current will match the _defaults except for when you are running tests and
+  // have swapped out for a stub/mock
+  let _current: Record<string, unknown> = {};
+  let _productionMode = false;
+
+  // get(name) -> T
+  // lookup by name and return _current.
+  // validates a default exists, as we'd only use this to lookup things that should exist
+  const get = <T>(name: string): T => {
+    if (_current[name] === undefined || _current[name] === undefined) {
+      throw new Error(`ioc.get("${name}") no current or default defined.`);
+    }
+    return _current[name] as T;
+  };
+
+  // setDefault(name, T)
+  // register a name for T
+  // names must be unique, and once set are permanent
+  // state in T is allowed and preserved across set/reset
+  // returns a iocGet wrapper for the name.
+  const setDefault = <T>(thing: { [name: string]: T }) => {
+    const entries = Object.entries(thing);
+    if (entries.length !== 1) {
+      throw new Error(`ioc.setDefault({name}) takes only one thing at a time.`);
+    }
+    const [name, t] = entries[0];
+    if (_defaults[name] !== undefined) {
+      throw new Error(`ioc.setDefault({${name}:...}) -> already set.}`);
+    }
+    _defaults[name] = t;
+    _current[name] = t;
+
+    return () => get<T>(name);
+  };
```

Comment:
Is there any reason why the signature of `setDefault` cannot be closer to the signature of `set`, and simply take in two params?

Suppose we did `const setDefault = <T>(name: string, t: T) => {`:
This would get rid of the need for the 'takes only one thing' error and make the API simpler.